### PR TITLE
Add `rome_formatter` fork as `ruff_formatter`

### PR DESCRIPTION
**Pull request by charliermarsh**
The Ruff autoformatter is going to be based on an intermediate representation (IR) formatted via [Wadler's algorithm](https://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf). This is architecturally similar to [Rome](https://github.com/rome/tools), Prettier, [Skip](https://github.com/skiplang/skip/blob/master/src/tools/printer/printer.sk), and others.

This PR adds a fork of the `rome_formatter` crate from [Rome](https://github.com/rome/tools), renamed here to `ruff_formatter`, which provides generic definitions for a formatter IR as well as a generic IR printer. (We've also pulled in `rome_rowan`, `rome_text_size`, and `rome_text_edit`, though some of these will be removed in future PRs.)

Why fork? `rome_formatter` contains code that's specific to Rome's AST representation (e.g., it relies on a fork of rust-analyzer's `rowan`), and we'll likely want to support different abstractions and formatting capabilities (there are already a few changes coming in future PRs). Once we've dropped `ruff_rowan` and trimmed down `ruff_formatter` to the code we currently need, it's also not a huge surface area to maintain and update.

### CODE REVIEW

**Comment by charliermarsh** [comment]
\cc @MichaReiser

File: crates/ruff_formatter/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+[package]
+name = "ruff_formatter"
+version = "0.0.0"
+edition = "2021"
+authors = ["Rome Tools Developers and Contributors"]
```
Comment (MichaReiser) [inline crates/ruff_formatter/Cargo.toml:5]: You may want to change the authors to Ruff.

**Review by MichaReiser — COMMENTED** [review]
LGTM

Some things to consider:

* Should we re-license the crates 
* Can we document the use of prior work in an acknowledgment section (pylint, Rome, Prettier...)
* It's probably worth unifying the logging infrastructure. Rome uses `tracing` whereas Ruff uses some other crates.

File: crates/ruff_formatter/Cargo.toml (hunk 1 above)
Comment (charliermarsh) [reply crates/ruff_formatter/Cargo.toml:5]: I wasn't certain what the "right" pattern was here. I noticed (e.g.) that `rome_rowan` has `authors = ["Aleksey Kladov <aleksey.kladov@gmail.com>"]`.

**Comment by charliermarsh** [comment]
> Should we re-license the crates

Ah yeah. How should this work? We already include a bunch of licenses in our `LICENSE` file -- could we add the existing licenses there, and then cover these forks under the project's MIT license?

> Can we document the use of prior work in an acknowledgment section (pylint, Rome, Prettier...)

Good idea -- I'll do this in a separate PR.

> It's probably worth unifying the logging infrastructure. Rome uses tracing whereas Ruff uses some other crates.

Also a good idea -- I'll also do this in a separate PR.

**Comment by MichaReiser** [comment]
> Ah yeah. How should this work? We already include a bunch of licenses in our LICENSE file -- could we add the existing licenses there, and then cover these forks under the project's MIT license?

I'm not really familiar with re-licensing but adding another copyright seems to work (if that's even something we want)

https://gist.github.com/fbaierl/1d740a7925a6e0e608824eb27a429370

**Pull request merged**